File: java-sdks/src/main/java/com/suisdks/sui/keypairs/Ed25519Keypair.java
package com.suisdks.sui.keypairs;

import com.suisdks.sui.cryptography.BouncyCastleProviderHolder;
import com.suisdks.sui.cryptography.SignatureScheme;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.util.Arrays;

public final class Ed25519Keypair implements Keypair {
    private final PrivateKey privateKey;
    private final PublicKey publicKey;

    private Ed25519Keypair(PrivateKey privateKey, PublicKey publicKey) {
        this.privateKey = privateKey;
        this.publicKey = publicKey;
    }

    public static Ed25519Keypair generate() {
        try {
            BouncyCastleProviderHolder.ensureInstalled();
            KeyPairGenerator g = KeyPairGenerator.getInstance("Ed25519", "BC");
            KeyPair kp = g.generateKeyPair();
            return new Ed25519Keypair(kp.getPrivate(), kp.getPublic());
        } catch (Exception e) {
            throw new IllegalStateException("Failed to generate Ed25519 keypair", e);
        }
    }

    @Override
    public SignatureScheme scheme() {
        return SignatureScheme.ED25519;
    }

    @Override
    public byte[] publicKeyBytes() {
        byte[] x509 = publicKey.getEncoded();
        return Arrays.copyOfRange(x509, x509.length - 32, x509.length);
    }

    public static PublicKey publicKeyFromRaw(byte[] raw32) {
        if (raw32.length != 32) {
            throw new IllegalArgumentException("Ed25519 public key must be 32 bytes");
        }
        // RFC8410 SubjectPublicKeyInfo prefix for Ed25519.
        byte[] prefix = new byte[] {0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00};
        byte[] full = new byte[prefix.length + raw32.length];
        System.arraycopy(prefix, 0, full, 0, prefix.length);
        System.arraycopy(raw32, 0, full, prefix.length, raw32.length);
        try {
            BouncyCastleProviderHolder.ensureInstalled();
            return KeyFactory.getInstance("Ed25519", "BC").generatePublic(new X509EncodedKeySpec(full));
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid Ed25519 public key bytes", e);
        }
    }

    @Override
    public byte[] sign(byte[] message) {
        try {
            Signature s = Signature.getInstance("Ed25519", "BC");
            s.initSign(privateKey);
            s.update(message);
            return s.sign();
        } catch (Exception e) {
            throw new IllegalStateException("Failed to sign with Ed25519", e);
        }
    }

    @Override
    public boolean verify(byte[] message, byte[] signature) {
        return verifyWithPublicKey(publicKeyBytes(), message, signature);
    }

    public static boolean verifyWithPublicKey(byte[] publicKey, byte[] message, byte[] signature) {
        try {
            Signature s = Signature.getInstance("Ed25519", "BC");
            s.initVerify(publicKeyFromRaw(publicKey));
            s.update(message);
            return s.verify(signature);
        } catch (Exception e) {
            return false;
        }
    }
}

File: kotlin-sdks/src/main/kotlin/com/suisdks/sui/keypairs/Ed25519Keypair.kt
package com.suisdks.sui.keypairs

import com.suisdks.sui.cryptography.BouncyCastleProviderHolder
import com.suisdks.sui.cryptography.SignatureScheme
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec

class Ed25519Keypair private constructor(
    private val privateKey: PrivateKey,
    private val publicKey: PublicKey,
) : Keypair {
    companion object {
        private val ED25519_SPKI_PREFIX = byteArrayOf(
            0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00,
        )

        fun generate(): Ed25519Keypair {
            return try {
                BouncyCastleProviderHolder.ensureInstalled()
                val generator = KeyPairGenerator.getInstance("Ed25519", "BC")
                val pair = generator.generateKeyPair()
                Ed25519Keypair(pair.private, pair.public)
            } catch (e: Exception) {
                throw IllegalStateException("Failed to generate Ed25519 keypair", e)
            }
        }

        fun publicKeyFromRaw(raw32: ByteArray): PublicKey {
            require(raw32.size == 32) { "Ed25519 public key must be 32 bytes" }
            val full = ED25519_SPKI_PREFIX + raw32
            return try {
                BouncyCastleProviderHolder.ensureInstalled()
                KeyFactory.getInstance("Ed25519", "BC").generatePublic(X509EncodedKeySpec(full))
            } catch (e: Exception) {
                throw IllegalArgumentException("Invalid Ed25519 public key bytes", e)
            }
        }

        fun verifyWithPublicKey(publicKey: ByteArray, message: ByteArray, signature: ByteArray): Boolean {
            return try {
                BouncyCastleProviderHolder.ensureInstalled()
                val verifier = Signature.getInstance("Ed25519", "BC")
                verifier.initVerify(publicKeyFromRaw(publicKey))
                verifier.update(message)
                verifier.verify(signature)
            } catch (_: Exception) {
                false
            }
        }
    }

    override fun scheme(): SignatureScheme = SignatureScheme.ED25519

    override fun publicKeyBytes(): ByteArray {
        val x509 = publicKey.encoded
        return x509.sliceArray(x509.size - 32 until x509.size)
    }

    override fun sign(message: ByteArray): ByteArray {
        return try {
            BouncyCastleProviderHolder.ensureInstalled()
            val signer = Signature.getInstance("Ed25519", "BC")
            signer.initSign(privateKey)
            signer.update(message)
            signer.sign()
        } catch (e: Exception) {
            throw IllegalStateException("Failed to sign with Ed25519", e)
        }
    }

    override fun verify(message: ByteArray, signature: ByteArray): Boolean {
        return verifyWithPublicKey(publicKeyBytes(), message, signature)
    }
}
